mesonbuild/modules/python.py

```diff
@@ -0,0 +1,384 @@
+
+        return interpreter.func_shared_module(None, args, kwargs)
+
+    def dependency(self, interpreter, state, args, kwargs):
+        dep = PythonDependency(self, interpreter.environment, kwargs)
+        return interpreter.holderify(dep)
+
+    @permittedSnippetKwargs(['pure', 'subdir'])
+    def install_sources(self, interpreter, state, args, kwargs):
+        pure = kwargs.pop('pure', False)
+        if not isinstance(pure, bool):
+            raise InvalidArguments('"pure" argument must be a boolean.')
+
+        subdir = kwargs.pop('subdir', '')
+        if not isinstance(subdir, str):
+            raise InvalidArguments('"subdir" argument must be a string.')
+
+        if pure:
+            kwargs['install_dir'] = os.path.join(self.purelib_install_path, subdir)
+        else:
+            kwargs['install_dir'] = os.path.join(self.platlib_install_path, subdir)
+
+        return interpreter.func_install_data(None, args, kwargs)
+
+    @noPosargs
+    @permittedKwargs(['pure', 'subdir'])
+    def get_install_dir(self, node, args, kwargs):
+        pure = kwargs.pop('pure', False)
+        if not isinstance(pure, bool):
+            raise InvalidArguments('"pure" argument must be a boolean.')
+
+        subdir = kwargs.pop('subdir', '')
+        if not isinstance(subdir, str):
+            raise InvalidArguments('"subdir" argument must be a string.')
+
+        if pure:
+            res = os.path.join(self.purelib_install_path, subdir)
+        else:
+            res = os.path.join(self.platlib_install_path, subdir)
+
+        return ModuleReturnValue(res, [])
+
+    @noPosargs
+    @noKwargs
+    def language_version(self, node, args, kwargs):
+        return ModuleReturnValue(self.version, [])
+
+    @noPosargs
+    @noKwargs
+    def found(self, node, args, kwargs):
+        return ModuleReturnValue(True, [])
+
+    @noKwargs
+    def get_path(self, node, args, kwargs):
+        if len(args) != 1:
+            raise InvalidArguments('get_path takes exactly one positional argument.')
+        path_name = args[0]
+        if not isinstance(path_name, str):
+            raise InvalidArguments('get_path argument must be a string.')
+
+        path = self.paths.get(path_name)
+
+        if path is None:
+            raise InvalidArguments('{} is not a valid path name'.format(path_name))
+
+        return ModuleReturnValue(path, [])
+
+    @noKwargs
+    def get_variable(self, node, args, kwargs):
+        if len(args) != 1:
+            raise InvalidArguments('get_variable takes exactly one positional argument.')
+        var_name = args[0]
+        if not isinstance(var_name, str):
+            raise InvalidArguments('get_variable argument must be a string.')
+
+        var = self.variables.get(var_name)
+
+        if var is None:
+            raise InvalidArguments('{} is not a valid path name'.format(var_name))
+
+        return ModuleReturnValue(var, [])
+
+    def method_call(self, method_name, args, kwargs):
```

Comment:
shouldn't this use a white-list of allowed methods? Reflection with user controlled data might open some things we rather not have our users poke around in.

Reply:
This is taken from https://github.com/mesonbuild/meson/blob/master/mesonbuild/interpreter.py#L1223

Reply:
That code at least filters out private and dunter methods. Which is the bare minimum. I think a white-list is cleaner, esp as this is not a class deriving from `ExtensionModule` where it might be more obvious that any non private helper method would be exposed as interpreter command.

Reply:
I've just pushed something to address this, figured it didn't cost much :)